CMakeLists.txt

```diff
@@ -543,12 +543,10 @@ set_target_properties(torrent-rasterbar
 		SOVERSION ${SOVERSION}
 )
 
-if (cxx_std_14 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
-	target_compile_features(torrent-rasterbar PUBLIC cxx_std_14)
-	message(STATUS "Building in C++14 mode")
-else()
-	target_compile_features(torrent-rasterbar PUBLIC cxx_std_11)
-	message(STATUS "Building in C++11 mode")
+# make the CXX_STANDARD property public to ensure it makes it into the pkg-config file
+get_target_property(_std torrent-rasterbar CXX_STANDARD)
+if (${_std})
+	target_compile_features(torrent-rasterbar PUBLIC cxx_std_${_std})
```

Comment:
Let me quote the manual once again:
`CXX_STANDARD`

This property specifies the C++ standard whose features are **requested**
to build this target. 

If the value requested **does not result** in a compile flag being added for
the compiler in use, **a previous standard flag will be added instead**.  This
means that using:
```cmake
 set_property(TARGET tgt PROPERTY CXX_STANDARD 11)
```
with a compiler which does not support ``-std=gnu++11`` or an equivalent
flag will not result in an error or warning, but will instead add the
``-std=gnu++98`` flag if supported.  This "decay" behavior may be controlled
with the ``CXX_STANDARD_REQUIRED`` target property.
Additionally, the ``CXX_EXTENSIONS`` target property may be used to
control whether compiler-specific extensions are enabled on a per-target basis.

*end of the quote*

But `target_compile_features()` sets features that can't be ignored. In other words, you change here the behaviour of ` CMAKE_CXX_STANDARD` variable and `CXX_STANDARD` making them enforcing as if `CXX_STANDARD_REQUIRED` was always set to `true`. Was it your intention?

Reply:
> But target_compile_features() sets features that can't be ignored. In other words, you change here the behaviour of CMAKE_CXX_STANDARD variable and CXX_STANDARD making them enforcing as if CXX_STANDARD_REQUIRED was always set to true. Was it your intention?

I don't see how this code does that. If the target doesn't have a value for the `CXX_STANDARD` property, this won't do anything. Presumably, if the target *does* have a `CXX_STANDARD` property set, surely it's *after* this decay logic has happened already, right? How else was the property set on the target?

Reply:
> Presumably, if the target does have a CXX_STANDARD property set, surely it's after this decay logic has happened already, right?

I think no. The decay, as far as I understand, happens right when CMake compares the value of `CXX_STANDARD` with the list of standards supported by the compiler. If `CXX_STANDARD_REQUIRED` is not set, CMake silently drops the requirement and shows an error otherwise.

Reply:
> so, you're saying the *effective* `CXX_STANDARD` property is determined at some point later? 

Yes.

> Is there a way to query that property?

I'm not aware of that. But we can test if the requested standard is on the `CMAKE_CXX_COMPILE_FEATURES` list. If it is there, `CXX_STANDARD` will be used. And we can safely enforce it via compile features. If not, we need to downgrade as well. It can be done taking a sublist of `CMAKE_CXX_KNOWN_FEATURES` (those prefixed with cxx_std_) and selecting previous records. We would need `list(FILTER)`, `list(TRANSFORM)`, `list(GET)` and a for-loop. I could write a snippet for doing that, if you like (albeit in a few hours).

> Or do you have an idea for how to do this differently?

I believe the test above does what is needed here.

Reply:
https://gitlab.kitware.com/cmake/cmake/issues/18446

We can wait for their suggestions.

Reply:
perhaps simpler solution would be to check to see if `CMAKE_CXX_STANDARD_REQUIRED` is set to `ON`, and if it isn't, fail the build.

Reply:
Maybe we'd better pack this mess into a macro and hide in its own file.

Reply:
@zeule would you mind posting this as a new PR?

Reply:
OK

Reply:
#3377